### ISSUE TITLE
As a developer, I want documentation to help me learn how to contribute to the WRES codebase

### ISSUE THREAD

**Issue by HankHerr-NOAA** [NOAA-OWP/wres#365]
Feel free to reword the title as needed. As discussed last week, we need documentation to help guide folks in contributing to the WRES. At a minimum, it should be something high-level to serve as an entry point into understanding the code base. It may not need to be much more than that; its unclear at this time.

The current developer wiki only discusses editing wikis:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Developers

This ticket is about fleshing that out a bit more. A checklist has been added below for specific items to address.

Hank

==========

- [ ] Write system testing wiki using VLab wiki as starting point (look for executing gradle wiki)
- [ ] Write a composition of artifacts description. See VLab wiki About Versions for a starting point. Update it.
- [ ] Write a wiki abut use of a database. VLab has information about setting up a Postgres database. This will be useful for users, as well.

**Comment by HankHerr-NOAA** [comment]
We obviously need a bit more information for developers to feel comfortable diving into the WRES, but please review my latest version of the text in [the comment above](https://github.com/NOAA-OWP/wres/issues/365#issuecomment-2698918356). Is it good enough to replace the current README? 

Thanks,

Hank

**Comment by james-d-brown** [comment]
I have made some edits inband to that comment to save faffing around - looks good.

**Comment by james-d-brown** [comment]
By the way, I omitted the reference to COWRES, because this is invisible/private.

**Comment by HankHerr-NOAA** [comment]
Yeah, I wasn't sure about mentioning COWRES in the README intended for developers. It is there in the user documentation wiki, but only because it just seems important for users of the WRES to know that it exists and who is able to use it.

Thanks for the review! I plan to push the changes to the README.md later this morning, so, if anyone else has recommended changes, please let me know.

Hank

**Comment by HankHerr-NOAA** [comment]
The pull request is in and checks being performed. I'll merge once complete.

Hank

**Comment by HankHerr-NOAA** [comment]
Once the pull request is merged, and the landing page updates, I plan to keep this ticket open. There is more developer related content that we might want to add to the wiki. What is there is now is just enough to get a developer to build, unit test, and execute the software, and nothing more. As we identify additional wikis, we can either do the work within this ticket or relate the work to this ticket. We'll close this ticket once we are content with the level of developer documentation.

Moving this ticket to on-hold. Again, I'll merge the code once the PR checks complete. Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
I'm working on this before my appointment. I'm starting with migrating existing wikis from VLab to GitHub where appropriate,

Hank

**Comment by HankHerr-NOAA** [comment]
Note to self... The VLab wiki About WRES Version Numbers includes a discussion of the composition of software artifacts which might be interesting for developers. It would probably need to be updated, but I'm writing this comment as a reminder to do that and add it to the developer wiki.

Hank

**Comment by HankHerr-NOAA** [comment]
How low level should we make the developer wiki? For example, in VLab, with a wiki on appropriate exception handling (just look for Exception Handling). Is that level of documentation necessary?

Hank

**Comment by HankHerr-NOAA** [comment]
System testing data is currently available in Google Drive and no where else. Please let me know if I'm wrong about that. Hence, outside folks wanting to collaborate will not have access to the system testing data. 

Is there value in providing instructions on system testing and setting up system testing data in the GitHub wiki given only internal folks will have access to the data (for now)? I'm on the fence about this one.

Of course, if I'm wrong, and system testing is available somewhere (can't recall for sure if this was done when we moved to GitHub), then including it in a wiki is a no brainer.

Hank

**Comment by james-d-brown** [comment]
We almost certainly need to host that data somewhere, which means we almost certainly need to make it smaller, which is something we've wanted to do anyway, right?

In terms of the level of detail, if there is valuable information in the internal wiki, I would probably copy it across.

**Comment by HankHerr-NOAA** [comment]
Thanks for your feedback. I'll create a system testing wiki that has the main developer wiki as a parent and then port it over. I'll edit it appropriately, removing any NWC specific stuff, and write the test data setup portion generically enough that editing it for however we choose the share the data in the future is simple. In fact, if we can get the data small enough to all fit in `smalldata`, this problem goes away altogether.  

However, that will take more than the amount of time I have now, so I'll leave that for later. I'm going to start a checklist to keep track of what I haven't done yet.

Hank

**Comment by james-d-brown** [comment]
I think it will always be too large for `smalldata`, which is hosted in the code repo (tbh, it should be probably be separated out too, but I digress). However, it should be rendered small enough for hosting on gdrive or similar and for fast download, probably a few GiB at most, ideally much smaller.

**Comment by HankHerr-NOAA** [comment]
All:

Given your recommendation to be succinct in the developer documentation, I'd like to get a feel for what you all feel is succinct, but also enough information to get started. The wiki to look at this this one:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Developers

I copied over wikis from VLab that describe use of an IDE and version numbers. I feel that these section might be too verbose, but want other opinions. Please take a look and let me know what you think about the level of detail.

Thanks,

Hank

**Comment by james-d-brown** [comment]
It's a mix. Mostly good, but certain parts can be absolutely slashed like:

`About WRES version numbers`

No need to go on about what could've been done. Also, these are all part of the `build.gradle`, not something a developer needs to worry about. It requires no more than one-line description.

**Comment by HankHerr-NOAA** [comment]
Thanks. I'll take that out tomorrow.

I'll then add a system testing wiki for developers and a wiki explaining how to use a database that would benefit both users and developers. They will be most be copy-and-paste with edits to make them more succinct, but time will tell.

Whether I get to those additions tomorrow or next week depends on how long the reset of the work on my docket takes. Have a great evening,

Hank

**Comment by james-d-brown** [comment]
`-Dwres.useDatabase=true`

**Comment by james-d-brown** [comment]
To run in memory, I generally give it 8 GiB, but it can probably work with less. For a database, 512 MiB is enough. I note that the above command doesn't assign this, so it will fall back to the default in the `build.gradle`, which is about 2GiB IIRC.

**Comment by HankHerr-NOAA** [comment]
Thanks. 

What would be better for a developer? A run script or direct commands?

Part of me feels like being able to run the commands by hand might help the developer better understand what is happening. The other part of me would really rather not run that long command. 

Thoughts?

Hank

**Comment by james-d-brown** [comment]
Well, it's just a copy/paste. A script is another file. Another thing, I wouldn't provide a command that depends on a linux emulator being installed, and associated utilities.

**Comment by HankHerr-NOAA** [comment]
Make sense. So no script.

I'm going to try to create a copy-and-paste command. The issue is that its much easier to provide the command using environment variables storing versions, which can be automatically obtained from .zip file names. However, that automation uses `bash`. If I provide the full command, then it will not be copy-and-paste; it will be copy-and-paste-and-edit, trying to dig out what to edit looking at a command hundreds of characters long. 

If that is the best approach,  providing a raw command and telling them to edit it, I can do that. I'm just not, personally, a big fan of that. Then again, its not easy for me to put myself in a generic developer's shoes.

Hank

**Comment by HankHerr-NOAA** [comment]
Note that the command I'm piggybacking off of uses .zip files, as does the VLab wikis providing the command. If there is a simpler way to run it that uses a local install, and not zips, let me know. That would make it so that the version numbers don't need to be replaced.

Thanks,

Hank

**Comment by james-d-brown** [comment]
If the user wants to make it into a script, they can. I don't see an issue with using placeholders in the example command for which environment variables can be set, but I think the developer should insert the software versions they want to test each time because they will vary - environment variables are best for fixed things. Obviously, a build dir will typically have many versions available.

**Comment by james-d-brown** [comment]
Again, again, I wouldn't use things like `tee` to redirect. The log is written to the default log location.

**Comment by HankHerr-NOAA** [comment]
True. The more I think about it, the more I realize I shouldn't coddle the developers. I'll give a raw command and try to at least make it obvious where changes need to be made.

And understood about `tee`; I've removed that command already in my tests.

System tests failed, but I'm guessing it has to do with memory, as I said. Still optimizing the command to provide them,

Hank

**Comment by HankHerr-NOAA** [comment]
Confirmed that using default memory setting fails due to `OutOfMemory`. Trying 4 GBs,

Hank

**Comment by HankHerr-NOAA** [comment]
System testing is described in the wiki:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Developers#instructions-for-running-system-tests

Its based on commands I ran which resulted in the system tests failing for 652 and 703, so there might be something wrong. I have other tasks I need to complete today, but, if I have time, I'll revisit these instructions to try to figure out if there is a problem.  Otherwise, that will happen next week. 

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
My 652 and 703 errors may be related to recent changes to the names of netCDF files. As such, pulled the latest from master again to make sure I picked up on any benchmark changes, reinstalled the system tests, and am running the tests again.  

Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
Unfortunately, other work has blocked progress on this. I'll try to get the database wiki up tomorrow.

Hank

**Comment by HankHerr-NOAA** [comment]
A first cut at the data wik is here:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-using-a-database-in-a-WRES-standalone-execution

I would appreciate a review to,

1. Identify mistakes I've made.
2. Identify needed information that is omitted.

I'll continue my review of the VLab wikis looking, as well, to see if I've left out anything or messed something up. Thanks,

Hank

**Comment by HankHerr-NOAA** [comment]
All:

Are we still maintaining the diagrams in the VLab WRES Pairing Database Schema and Related Diagrams wiki? We have updating that wiki as a step in the deployment process, but there has been no change in 2022. However, that could just reflect the reality of how rarely we update the schema.  

If that wiki is still accurate, then I I'll add it as a development resource in GitHub and deprecate the VLab version.

Hank

**Comment by james-d-brown** [comment]
Probably shouldn't call it a "Pairing Database" because it really has nothing to do with pairing, it is just organizing time-series data and associated metadata. Since we use liquibase to manage the database, you can look at the `db.changelog-master.xml` for a git history, but it very rarely changes. Last time would have been declarative partitions, but that was under the hood.

**Comment by HankHerr-NOAA** [comment]
The `evaluation_id` was added to the `executionlog` in Jan, 2024, based on the commit comment, but that table is not covered in the diagram showing in the wiki. So I guess there was no reason to update the diagram, which is why its from 2022.

I'm going to create a developer wiki including the database diagram and other language from the VLab wiki, and include the query I often perform (which Evan has added to the monitoring tool) to obtain a history of the evaluations. Its a query that has served me well over the years, so it would probably be useful for potential developers, as well.

Hank

**Comment by HankHerr-NOAA** [comment]
A very rough beginning for a developer database wiki is here:

https://github.com/NOAA-OWP/wres/wiki/Developer's-Guide-to-the-Database

More tomorrow,

Hank

**Comment by HankHerr-NOAA** [comment]
Evan provided me this query borrowed from the monitoring tool:

```sql
select el.project_name, el.evaluation_id, substring(el.arguments,0,12) , el.failed, el.start_time, el.end_time, (extract (epoch from el.end_time) - extract (epoch from el.start_time)) / 60 as duration_minutes
from wres.executionlog el
where start_time > CURRENT_DATE - $DaysAgo
order by start_time;
```

I need to compare this with what I shared in the wiki (based on my own query history) and make appropriate changes. Again, more tomorrow,

Hank

**Comment by HankHerr-NOAA** [comment]
I made changes to the developer database wiki based on the query Evan shared. 

I'm removing myself as assignee until I can come back around to this task. I've gotten a start developer documentation, but have not completed my scan of the VLab wikis to determine if there are others to migrate which are at the appropriate level considering the audience. And with other work on the horizon, I may not be able to get get back to this work for another month or two or three.

Please feel free to review what I have written and make changes as you see fit. That includes:

https://github.com/NOAA-OWP/wres/wiki/Instructions-for-using-a-database-in-a-WRES-standalone-execution
https://github.com/NOAA-OWP/wres/wiki/Instructions-for-Developers
https://github.com/NOAA-OWP/wres/wiki/Developer%27s-Guide-to-the-Database

And, of course, the README. Thanks,

Hank
